### DISCUSSION TITLE
Some technical questions on lighthouse

### DISCUSSION THREAD

**Discussion by runger1101001** [bitcraze/discussions#478] [Q&A] [answered]
Dear community!

First off, a great thank you for the wonderful tech you've created, and the active community around it!

We're working an adaptation of the lighthouse deck for use on some ground-based robots, and I have a couple of questions I was hoping someone here could help me out with:

1. For the lighthouse deck geometry, is it correct to assume that the 15 x 30mm spacing of the photo-diodes was chosen based on the space available on the crazyflie, and one could also use different dimensions, and just adapt the values in the STM32-based driver code? Or are there other important reasons for the 15x30mm spacing?

2. Programming the Lattice FPGA - the way the rev D schematics of the lighthouse deck are set up, the Winbond flash chip, the FPGA and the SPI header all share the same SPI lines, including nCS. I must say that I have not understood the programming concept. Won't the flash chip interfere when trying to program the FPGA, since they share the same nCS? Or is the idea only to program the flash, while holding the RESET line of the FPGA high? Because I don't see how you can program the NVCM on the FPGA in this configuration. The IceBreaker demo board for this chip has a considerably more complex config for this, involving several solder-jumpers. If someone has some insights how this works on the lighthouse deck, or has pointers to relevant documentation, I'd be really glad to learn.

Of course my design adaptations will be open source, and once its halfway finished I would be very happy to share details here if anyone is interested!

**Comment by ataffanel** [comment, marked as answer]
Hi, interesting project!

I can answer in order:
 1. This is correct, we spaced them as much as possible to fit on the Crazyflie (and keep the spacing a "round" number ;-).
 2. A [UART/I2C bootloader](https://github.com/bitcraze/lighthouse-bootloader) if flashed in the SPI flash. The bootloader is used to then program the flash from the Crazyflie or any outside board. The ICE40 supports a "multiboot" mode where there can be multiple bitstream in the spi flash and we can launch the reconfiguration from within the fpga (very neat!). Your assumption is correct though, in factory we keep the FPGA reset to flash the multiboot partition table and the bootloader in SPI flash. The idea is that there is then no more need to directly touch the flash. All the required line to program the SPI flash are available behind open solder-bridge, this is what was used for the bootloader development.

One side note: the Ice40UP5k was chosen mostly because of physical size constrains, it was the biggest one available on qfn48 package. Though is has been causing problem because it is quite slow and small. If you have more space or feel OK with using BGA you could consider the 8K variants of the ICE40 FPGA.

**Reply by runger1101001** [reply]
Hey, thanks a lot for your detailed reply!

I think I'll stick to the same FPGA for now to not have too many different parts. I'll post here when we get something working!
